### PR TITLE
H286: Aggregation operator sweep on H275 SOTA — untested reduction axis

### PR DESCRIPTION
**Pull request by morganmcg1**
# H286: Aggregation operator sweep on H275 SOTA recipe — untested reduction axis

## Hypothesis

**The H275 SOTA recipe uses mean aggregation over 36 noise-perturbed predictions per case. Mean is BLUE only under Gaussian residuals; if predictions have heavy tails (some perturbations land in bad valleys), robust aggregators could improve.**

H275 SOTA (val 5.9243 / test 5.7690 / test_WSS 6.6743) reduces 36 raw predictions per (case, channel) — K=3 pairs × 2 anti × 6 res × 2 mirror — via simple arithmetic mean. **No one has tested whether a different aggregator gives a free win on the same data.**

Three concrete predictions:
1. **If predictions are clean Gaussian** → all aggregators give ~same answer, mean wins by a hair. We bank Finding CCC-mean-is-optimal-for-stacked-anti and free compute for non-aggregation axes.
2. **If some perturbations land in bad noise valleys** → trimmed mean (10-20%) or median wins on noisy channels (test_WSS especially — Morgan's #1 metric, most heterogeneous channel).
3. **If residuals are heavy-tailed and asymmetric** → Huber M-estimator wins, possibly +3-5bp on test_WSS.

**Predicted val_abupt: 5.918-5.928** (-6bp to +5bp vs H275 5.9243, depending on which aggregator wins).

This is the highest-EV experiment available right now because:
- **Zero extra compute**: ONE eval pass produces predictions for all aggregators (gather per-pass outputs, apply each reduction post-hoc)
- **Untested mechanism axis**: K, σ, EP, mirror, resolution, antithetic, Sobol-direction are all explored; reduction operator is not
- **Targets test_WSS specifically**: WSS has the largest channel-to-channel variance in raw predictions — most likely to benefit from robust averaging

## Why this is the right next step for you

- You just executed a clean K=2 anti-thetic eval (Finding BBB-K-curve-floor-EP13 from H282) — you know `eval_tta_h252.py` intimately
- Your K-curve analysis showed the linear cancellation mechanism dominates K-averaging variance reduction at K=3 — this naturally raises the question of what reduction to apply
- You're the natural fit for clean methodical aggregator sweeps
- Same engine, same checkpoint, same K=3 anti — only the final reduction step changes

## Why complementary to in-flight work

| Recipe | Axis | What it tests |
|---|---|---|
| H275 SOTA (merged) | baseline | EP15 × anti-K=3 mean |
| H285 tanjiro (in-flight) | K-scaling | K=4 at EP15 |
| H284 edward (in-flight) | distribution | Sobol-anti K=3 at EP15 |
| H283 thorfinn (in-flight) | σ × Sobol | σ=3e-4 + Sobol-K=5 at EP15 |
| **H286 (you)** | **aggregation** | **mean vs trimmed/median/Huber at EP15** |

Aggregation is orthogonal to K, σ, distribution, and checkpoint. A win here STACKS on top of any in-flight winner.

### Step 1: Resolve any merge conflicts from H275 SOTA merge

```bash
git fetch origin tay
git checkout -b alphonse/h286-aggregation-sweep origin/tay
```

### Step 2: Patch eval_tta_h252.py to gather per-pass predictions and add aggregator CLI

The script currently mean-reduces inside the eval loop. We need to:
1. **Cache the raw per-pass predictions** in a tensor of shape `[n_passes, n_points, n_channels]` per case
2. **Apply multiple aggregators after the loop** and log all of them to W&B in one run

**Recommended minimal patch:**

#### Step 2a: In the eval inner loop, store raw predictions

Find the section in `eval_tta_h252.py` that accumulates predictions per (res, mirror, k_pass) — currently it likely does running-mean. Change it to store each prediction:

```python
# replace running-mean accumulator with raw collection
raw_preds = []  # list of tensors, each [n_points, n_channels]
for res_idx, res in enumerate(resolutions):
    for mirror_idx, mirror in enumerate([False, True]):
        for k_idx in range(weight_noise_passes * (2 if antithetic_noise else 1)):
            pred = forward(case, res, mirror, k_idx)
            raw_preds.append(pred)  # store, don't accumulate
stacked = torch.stack(raw_preds, dim=0)  # [n_passes, n_points, n_channels]
```

#### Step 2b: Add aggregator functions

```python
def aggregate(stacked, mode, trim_frac=0.1, huber_k=1.345):
    """stacked: [n_passes, n_points, n_channels]"""
    if mode == "mean":
        return stacked.mean(dim=0)
    elif mode == "median":
        return stacked.median(dim=0).values
    elif mode == "trimmed_mean":
        # sort along pass dim, drop trim_frac fraction from each tail
        sorted_, _ = stacked.sort(dim=0)
        n = stacked.shape[0]
        k = int(n * trim_frac)
        if k > 0:
            sorted_ = sorted_[k:n-k]
        return sorted_.mean(dim=0)
    elif mode == "huber":
        # iterative reweighted least squares — converge in 3-5 iters
        med = stacked.median(dim=0).values
        for _ in range(5):
            residuals = stacked - med.unsqueeze(0)
            mad = residuals.abs().median(dim=0).values
            sigma = (mad / 0.6745).clamp(min=1e-8)
            normalized = residuals / sigma.unsqueeze(0)
            w = torch.where(normalized.abs() < huber_k, torch.ones_like(normalized), huber_k / normalized.abs())
            med = (w * stacked).sum(dim=0) / w.sum(dim=0).clamp(min=1e-8)
        return med
    elif mode == "geometric_mean":
        # only for positive predictions — guard with sign
        signs = stacked.sign().prod(dim=0)
        return signs * stacked.abs().clamp(min=1e-12).log().mean(dim=0).exp()
    else:
        raise ValueError(f"Unknown aggregator: {mode}")
```

#### Step 2c: Apply all aggregators and log each

```python
aggregators = ["mean", "median", "trimmed_mean_10", "trimmed_mean_20", "huber"]
for agg in aggregators:
    if agg.startswith("trimmed_mean_"):
        frac = int(agg.split("_")[-1]) / 100.0
        reduced = aggregate(stacked, "trimmed_mean", trim_frac=frac)
    else:
        reduced = aggregate(stacked, agg)
    # compute per-channel rel_l2_pct vs ground truth
    metrics = compute_metrics(reduced, target)
    for ch, val in metrics.items():
        wandb.log({f"{agg}/{ch}_rel_l2_pct": val, "agg_mode": agg}, commit=False)
wandb.log({}, commit=True)
```

This gives you 5 aggregators × 7 channels = 35 metrics per (split, case) — log per-split aggregates with prefix `{split}_{agg}/`.

### Step 3: Verify EP15 checkpoint

```bash
H185_EP15_CKPT="outputs/drivaerml/run-0gjfv45i/checkpoint_ep15.pt"
# OR (byte-identical alternative path):
# H185_EP15_CKPT="outputs/h252_eval/_artifacts/0gjfv45i/epoch-15/checkpoint.pt"
ls -lh $H185_EP15_CKPT
```

### Step 4: Run primary command (same H275 recipe, multi-aggregator output)

```bash
torchrun --standalone --nproc-per-node=8 eval_tta_h252.py \
  --checkpoint $H185_EP15_CKPT \
  --resolutions "32768,49152,65536,81920,98304,131072" \
  --eval-modes "weight_noise_mirror_res_avg" \
  --weight-noise-sigma 5e-4 \
  --weight-noise-passes 3 \
  --antithetic-noise \
  --aggregators "mean,median,trimmed_mean_10,trimmed_mean_20,huber" \
  --batch-size 2 --num-workers 4 \
  --agent alphonse \
  --wandb-group "h286-alphonse-aggregation-sweep" \
  --wandb-name "alphonse/h286-ep15-aggregation-sweep"
```

Note: confirmed flag names from `eval_tta_h252.py` source (Finding from H282):
- `--weight-noise-passes 3` (NOT `--n-weight-noise-passes`)
- `--antithetic-noise` (NOT `--antithetic`)
- `--eval-modes "weight_noise_mirror_res_avg"` (exact mode string)

You'll need to add the `--aggregators` flag in your patch.

### Step 5: Defensive NCCL timeout patch

Frieren H279 and thorfinn H283 both crashed with NCCL allgather timeout (uneven val split: 36 cases / 8 ranks → ranks 0/1 each get 5 cases vs ranks 2-7 each get 4; 600s default tips over). Their fix in `eval_h271_sobol_tta.py:1049` was `timeout=timedelta(minutes=60)` in `init_distributed`.

`eval_tta_h252.py` may have the same issue. Before launch, search for the equivalent `init_distributed(...)` call and add `timeout=timedelta(minutes=60)`. Same compute budget as H275 (~230 min), low NCCL risk, but the patch is a one-liner defensive measure.

### Step 6: Post terminal SENPAI-RESULT with the aggregator sweep table

```
SENPAI-RESULT: {"terminal":true,"status":"complete","pending_arms":false,"wandb_run_ids":["<run-id>"],"primary_metric":{"name":"val_abupt_ep15_anti_K3_best_agg","value":<val>},"test_metric":{"name":"test_abupt_ep15_anti_K3_best_agg","value":<test>}}
```

Include the aggregator sweep table:

| Aggregator | val_abupt | test_abupt | test_WSS | test_SP | test_VP | Δ vs mean |
|---|---:|---:|---:|---:|---:|---:|
| mean (≡ H275 SOTA) | 5.9243 | 5.7690 | 6.6743 | 3.6427 | 3.3788 | reference |
| median | ? | ? | ? | ? | ? | Δ |
| trimmed_mean_10 | ? | ? | ? | ? | ? | Δ |
| trimmed_mean_20 | ? | ? | ? | ? | ? | Δ |
| huber | ? | ? | ? | ? | ? | Δ |
| **best (this)** | ? | ? | ? | ? | ? | **Δ** |

Plus 2-3 sentences: which aggregator wins on which channel? Is the pattern (mean for SP/VP, robust for WSS) the expected one? Does the **best per-channel aggregator** (mix-and-match) give an additional gain?

**If best_aggregator val < 5.9243 AND test < 5.7690 → immediate SOTA merge**

## Baseline metrics

| Reference | val_abupt | test_abupt | test_WSS | test_VP | test_SP | W&B |
|---|---:|---:|---:|---:|---:|---|
| **H275 EP15+anti-K=3+6-res+mirror ← CURRENT SOTA** | **5.9243** | **5.7690** | **6.6743** | **3.3788** | **3.6427** | 0b4t2bz2 |
| H274 EP13+anti-K=3+6-res+mirror | 5.9322 | 5.7763 | 6.6805 | 3.3846 | 3.6515 | o8oq9r92 |
| H278 EP13+anti-K=4+6-res+mirror (your peer) | 5.9316 | 5.7758 | 6.6798 | 3.3850 | 3.6508 | ardsfopx |
| H282 (your) EP13+anti-K=2+6-res+mirror | 5.9340 | 5.7783 | 6.6827 | 3.3867 | 3.6511 | 8jlfgz8y |

**Merge gate**: val_abupt < **5.9243** AND test_abupt < **5.7690**

## Critical constraints

- **DDP 8 GPUs** (`--nproc-per-node=8`) for every eval run
- **Read-only**: data/loader.py, data/preload.py, data/split_manifest.json, model code (eval-time TTA only)
- **No model capacity changes** — only the post-eval aggregation step changes
- **z-axis tau_z LOCKED at 1.67**
- **SENPAI_TIMEOUT_MINUTES=360** (~230 min predicted, same as H275)

## Memory note

Storing per-pass predictions for all 36 passes × 50k points × 7 channels × bf16 ≈ 12MB per case × 50 test cases ≈ 600 MB total — well within memory budget. Move to CPU between cases if needed.

## Suggested follow-ups (post-result)

1. If a single aggregator wins on all channels → bank "Finding CCC-{agg}-wins-stacked-anti" and assign K=5 anti-K=3 EP15 with {agg} (new SOTA candidate)
2. If different aggregators win on different channels → assign **per-channel best aggregator** at K=3 EP15 — could give an additional 1-3bp test_WSS gain
3. If mean is optimal across all → bank Finding "mean is BLUE for stacked anti at SOTA recipe", focus future TTA work on per-resolution weighting or asymmetric mirror-flip ensembling
4. Aggregator × σ interaction: if trimmed mean wins, σ might be retunable (Finding RR/XX) — trimmed mean tolerates larger σ

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT alphonse: Picked up PR. Found the H286 patch was pre-applied to eval_tta_h252.py. **Bug detected and fixed**: the original patch placed the DataLoader-iteration loop INSIDE the mirror-flag loop, causing the loader to be re-spawned per mirror (doubling worker spawn / data-load cost — confirmed by debug timing: 686.8s/case vs 453.2s/case after fix). Restructured so the loader iterates once per (k, sign, R) and accumulates to per-mirror pass tensors. Both passes (orig + mirror) are still collected as separate slices of the [n_passes, n_points, n_channels] stack, so aggregator semantics are unchanged.

Now launching the full sweep — H275 SOTA recipe + 5 aggregators (mean, median, trimmed_mean_10, trimmed_mean_20, huber). Expected ~210 min.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #1471 because ADVISOR: Closing H286 — all 5 aggregation operators fail the val gate (< 5.9235%).

**Finding OOO: Aggregation operator is a NULL axis for SOTA improvement on EP15+anti-K3+6-res stack.**

W&B run bad1i1jv (alphonse/h286-ep15-aggregation-sweep) complete val results:

| Aggregator | val_abupt | Gate (< 5.9235%) |
|---|---:|:---|
| trimmed_mean_10 | 5.9242% | FAIL (+0.07bp) |
| mean | 5.9243% | FAIL (+0.08bp) |
| trimmed_mean_20 | 5.9258% | FAIL (+0.23bp) |
| huber | 5.9261% | FAIL (+0.26bp) |
| median | 5.9303% | FAIL (+0.68bp) |

The baseline mean aggregator (5.9243%) is already near-optimal for this recipe. Robust alternatives (trimmed mean, huber, median) all add variance rather than reduce error. The standard mean remains the correct choice.

No test arms needed — no aggregator clears the val gate.

Good work on the bug fix (DataLoader restructuring) and thorough sweep. This definitively closes the aggregation operator axis. The test-time compute budget is better spent on the K-axis and resolution-axis experiments already in flight..
